pkg/kube/multicluster/secretcontroller.go

```diff
@@ -406,33 +406,24 @@ func (c *Controller) deleteSecret(secretKey string) {
 			log.Infof("ignoring delete cluster %v from secret %v as it would overwrite the config cluster", c.configClusterID, secretKey)
 			continue
 		}
-		log.Infof("Deleting cluster_id=%v configured by secret=%v", cluster.ID, secretKey)
-		cluster.Stop()
-		err := c.handleDelete(cluster.ID)
-		if err != nil {
-			log.Errorf("Error removing cluster_id=%v configured by secret=%v: %v",
-				cluster.ID, secretKey, err)
-		}
-		c.cs.Delete(secretKey, cluster.ID)
+
+		c.deleteCluster(secretKey, cluster)
 	}
 
 	log.Infof("Number of remote clusters: %d", c.cs.Len())
 }
 
-func (c *Controller) deleteCluster(secretKey string, clusterID cluster.ID) {
-	c.cs.Lock()
-	defer func() {
-		c.cs.Unlock()
-		log.Infof("Number of remote clusters: %d", c.cs.Len())
-	}()
-	log.Infof("Deleting cluster_id=%v configured by secret=%v", clusterID, secretKey)
-	c.cs.remoteClusters[secretKey][clusterID].Stop()
-	err := c.handleDelete(clusterID)
+func (c *Controller) deleteCluster(secretKey string, cluster *Cluster) {
+	log.Infof("Deleting cluster_id=%v configured by secret=%v", cluster.ID, secretKey)
+	cluster.Stop()
+	err := c.handleDelete(cluster.ID)
 	if err != nil {
 		log.Errorf("Error removing cluster_id=%v configured by secret=%v: %v",
-			clusterID, secretKey, err)
+			cluster.ID, secretKey, err)
 	}
-	delete(c.cs.remoteClusters[secretKey], clusterID)
+	c.cs.Delete(secretKey, cluster.ID)
```

Comment:
This is good, we should call the method of store instead of r/w its fields directly

Reply:
@hzxuzhonghu @SpecialYang - the logic changes LGTM especially WRT to the dirty data - the test changes concern me and since they appear to be passing point to possibly another logic flaw

Reply:
The test cases should be more explicit